pytorch_lightning/trainer/supporters.py

```diff
@@ -176,3 +181,316 @@ def to_disk(self) -> None:
+
+        finally:
+            self.counter += 1
+
+    def __len__(self) -> Union[int, float]:
+        return self.length
+
+
+class CombinedDataset(object):
+    """
+    Combine multiple datasets and compute their statistics
+    """
+    def __init__(self, datasets: Union[Sequence, Mapping]):
+        """
+
+        Args:
+            datasets: a sequence/mapping datasets. Can be a collections of torch.utils.Dataset,
+                Iterable or even None.
+
+        """
+        self.datasets = datasets
+
+        self.max_len = self._calc_num_data(self.datasets, 'max')
+        self.min_len = self._calc_num_data(self.datasets, 'min')
+
+    @staticmethod
+    def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str = 'min') -> Union[int, float]:
+        """
+        Compute the length of `CombinedDataset` according to the `mode`.
+
+        Args:
+            datasets: a sequence/mapping datasets. Can be a collections of torch.utils.data.Dataset,
+                Iterable or even None.
+            mode: 'min' or 'max'. Determine `CombinedDataset`'s length is the maximum or minimum of
+                the datasets.
+
+        Returns:
+            length: the length of `CombinedDataset`
+
+        """
+        if mode not in ['min', 'max']:
+            raise ValueError(f"Invalid Mode: {mode}")
+
+        # extract the lengths
+        all_lengths = apply_to_collection(datasets, (Dataset, Iterable, type(None)), get_len,
+                                          wrong_dtype=(Sequence, Mapping))
+
+        compute_func = eval(mode)
+
+        if isinstance(all_lengths, (int, float)):
+            length = all_lengths
+
+        elif isinstance(all_lengths, Mapping):
+            length = compute_func(all_lengths.values())
+
+        elif isinstance(all_lengths, Sequence):
+            length = compute_func(all_lengths)
+
+        return length
+
+    def __len__(self) -> int:
+        """Return the minimum length of the datasets."""
+        return self.min_len
+
+
+class CombinedLoader(object):
+    """
+    Combines different dataloaders and allows sampling in parallel.
+
+    Examples:
+        >>> loaders = {'a': torch.utils.data.DataLoader(range(6), batch_size=4),
+        ...            'b': torch.utils.data.DataLoader(range(15), batch_size=5)}
+        >>> combined_loader = CombinedLoader(loaders, 'max_size_cycle')
+        >>> for item in combined_loader:
+        ...     print(item)
+        {'a': tensor([0, 1, 2, 3]), 'b': tensor([0, 1, 2, 3, 4])}
+        {'a': tensor([4, 5]), 'b': tensor([5, 6, 7, 8, 9])}
+        {'a': tensor([0, 1, 2, 3]), 'b': tensor([10, 11, 12, 13, 14])}
+        >>> combined_loader = CombinedLoader(loaders, 'min_size')
+        >>> for item in combined_loader:
+        ...     print(item)
+        {'a': tensor([0, 1, 2, 3]), 'b': tensor([0, 1, 2, 3, 4])}
+        {'a': tensor([4, 5]), 'b': tensor([5, 6, 7, 8, 9])}
+
+    """
+    SUPPORTED_MODES = ('min_size', 'max_size_cycle')
```

Comment:
Can you add an explanation on what those modes are ?

Reply:
done